File: src/csharp/benchmarks/AgentEndToEnd/Fixtures/HighAssurance.Postgres/Execution/PostgresAuthorizationContextStore.cs
using Foundgine.Runtime.ControlPlane;
using Npgsql;

namespace Foundgine.HighAssurance.Postgres.Execution;

/// <summary>
/// PostgreSQL-backed authorization evidence whose row lock is held until the
/// surrounding mutation transaction commits or rolls back.
///
/// Authorization lifecycle writes additionally require an explicitly registered
/// writer provenance. The writer is bound to a tenant, actor, PostgreSQL role,
/// and monotonically increasing write sequence.
/// </summary>
public sealed class PostgresAuthorizationContextStore
{
    private readonly AuthorizationContextIntegrityKeyRingManager _integrityManager;
    private AuthorizationContextIntegrityKeyRing Integrity => _integrityManager.Snapshot;

    public PostgresAuthorizationContextStore(AuthorizationContextIntegrityKeyRing integrity)
    {
        ArgumentNullException.ThrowIfNull(integrity);
        _integrityManager = new AuthorizationContextIntegrityKeyRingManager(integrity);
    }

    public PostgresAuthorizationContextStore(AuthorizationContextIntegrityKeyRingManager integrityManager)
    {
        _integrityManager = integrityManager ?? throw new ArgumentNullException(nameof(integrityManager));
    }

    public AuthorizationContextIntegrityKeyRing CurrentIntegrityKeyRing => Integrity;

    /// <summary>
    /// Returns the integrity key ids still referenced by live authorization evidence
    /// or lifecycle tombstones. Call this inside the same administrative transaction
    /// that coordinates a retirement decision.
    /// </summary>
    public async Task<IReadOnlySet<string>> GetPersistedIntegrityKeyIdsAsync(
        NpgsqlConnection connection,
        NpgsqlTransaction transaction,
        CancellationToken cancellationToken = default)
    {
        const string sql = """
                           SELECT integrity_key_id FROM banking.authorization_context
                           UNION
                           SELECT integrity_key_id FROM banking.authorization_context_tombstone;
                           """;
        var result = new HashSet<string>(StringComparer.Ordinal);
        await using var command = new NpgsqlCommand(sql, connection, transaction);
        await using var reader = await command.ExecuteReaderAsync(cancellationToken);
        while (await reader.ReadAsync(cancellationToken))
            result.Add(reader.GetString(0));
        return result;
    }


    public async Task<AuthorizationContextRow?> LoadForUpdateAsync(
        NpgsqlConnection connection,
        NpgsqlTransaction transaction,
        Guid actorId,
        int tenantId,
        CancellationToken cancellationToken = default)
    {
        const string sql = """
                           SELECT actor_id, tenant_id, allowed, version, fingerprint, integrity_algorithm, integrity_key_id, integrity_tag
                           FROM banking.authorization_context
                           WHERE actor_id = @actor AND tenant_id = @tenant
                           FOR UPDATE;
                           """;

        await using var command = new NpgsqlCommand(sql, connection, transaction);
        command.Parameters.AddWithValue("actor", actorId);
        command.Parameters.AddWithValue("tenant", tenantId);
        await using var reader = await command.ExecuteReaderAsync(cancellationToken);
        if (!await reader.ReadAsync(cancellationToken))
            return null;

        var row = new AuthorizationContextRow(
            reader.GetGuid(0), reader.GetInt32(1), reader.GetBoolean(2),
            reader.GetInt64(3), reader.GetString(4), reader.GetString(5), reader.GetString(6), reader.GetString(7));
        if (!Integrity.VerifyContextTag(row.ActorId, row.TenantId, row.Allowed, row.Version, row.Fingerprint,
                row.IntegrityAlgorithm, row.IntegrityKeyId, row.IntegrityTag))
            throw new InvalidOperationException(
                "Authorization context integrity verification failed; authorization fails closed.");
        return row;
    }

    public Task CreateAsync(
        NpgsqlConnection connection, NpgsqlTransaction transaction,
        Guid actorId, int tenantId, bool allowed, long version, string fingerprint,
        CancellationToken cancellationToken = default) =>
        CreateAsync(connection, transaction, actorId, tenantId, allowed, version, fingerprint,
            new AuthorizationWriteProvenance(Guid.Empty, actorId, tenantId, 0, ""), cancellationToken);

    public async Task CreateAsync(
        NpgsqlConnection connection, NpgsqlTransaction transaction,
        Guid actorId, int tenantId, bool allowed, long version, string fingerprint,
        AuthorizationWriteProvenance provenance, CancellationToken cancellationToken = default)
    {
        ValidateVersion(version);
        ValidateFingerprint(fingerprint);
        await ValidateWriterAsync(connection, transaction, provenance, actorId, tenantId, cancellationToken);

        const string tombstoneSql = """
                                    SELECT last_version, last_fingerprint, integrity_algorithm, integrity_key_id, integrity_tag FROM banking.authorization_context_tombstone
                                    WHERE actor_id = @actor AND tenant_id = @tenant FOR UPDATE;
                                    """;
        await using (var tombstoneCommand = new NpgsqlCommand(tombstoneSql, connection, transaction))
        {
            tombstoneCommand.Parameters.AddWithValue("actor", actorId);
            tombstoneCommand.Parameters.AddWithValue("tenant", tenantId);
            await using var tombstoneReader = await tombstoneCommand.ExecuteReaderAsync(cancellationToken);
            if (await tombstoneReader.ReadAsync(cancellationToken))
            {
                var lastVersion = tombstoneReader.GetInt64(0);
                var lastFingerprint = tombstoneReader.GetString(1);
                var algorithm = tombstoneReader.GetString(2);
                var keyId = tombstoneReader.GetString(3);
                var tag = tombstoneReader.GetString(4);
                if (!Integrity.VerifyTombstoneTag(actorId, tenantId, lastVersion, lastFingerprint, algorithm, keyId,
                        tag))
                    throw new InvalidOperationException(
                        "Authorization lifecycle tombstone integrity verification failed; authorization fails closed.");
                if (version <= lastVersion)
                    throw new InvalidOperationException(
                        $"Authorization context version must exceed the last committed lifecycle version. Last={lastVersion}, requested={version}.");
            }
        }

        const string sql = """
                           INSERT INTO banking.authorization_context(actor_id, tenant_id, allowed, version, fingerprint, integrity_algorithm, integrity_key_id, integrity_tag)
                           VALUES (@actor, @tenant, @allowed, @version, @fingerprint, @algorithm, @key_id, @tag);
                           """;
        await using var command = new NpgsqlCommand(sql, connection, transaction);
        AddParameters(command, actorId, tenantId, allowed, version, fingerprint);
        command.Parameters.AddWithValue("algorithm", AuthorizationContextIntegrityKeyRing.CurrentAlgorithmVersion);
        command.Parameters.AddWithValue("key_id", Integrity.ActiveKeyId);
        command.Parameters.AddWithValue("tag",
            Integrity.ComputeContextTag(actorId, tenantId, allowed, version, fingerprint));
        await command.ExecuteNonQueryAsync(cancellationToken);
    }

    public Task UpdateAsync(
        NpgsqlConnection connection, NpgsqlTransaction transaction,
        Guid actorId, int tenantId, bool allowed, long version, string fingerprint,
        CancellationToken cancellationToken = default) =>
        UpdateAsync(connection, transaction, actorId, tenantId, allowed, version, fingerprint,
            new AuthorizationWriteProvenance(Guid.Empty, actorId, tenantId, 0, ""), cancellationToken);

    public async Task UpdateAsync(
        NpgsqlConnection connection, NpgsqlTransaction transaction,
        Guid actorId, int tenantId, bool allowed, long version, string fingerprint,
        AuthorizationWriteProvenance provenance, CancellationToken cancellationToken = default)
    {
        ValidateVersion(version);
        ValidateFingerprint(fingerprint);
        await ValidateWriterAsync(connection, transaction, provenance, actorId, tenantId, cancellationToken);

        var current = await LoadForUpdateAsync(connection, transaction, actorId, tenantId, cancellationToken)
                      ?? throw new InvalidOperationException("Authorization context does not exist.");
        if (version <= current.Version)
            throw new InvalidOperationException(
                $"Authorization context version must increase monotonically. Current={current.Version}, requested={version}.");

        const string sql = """
                           UPDATE banking.authorization_context
                           SET allowed = @allowed, version = @version, fingerprint = @fingerprint,
                               integrity_algorithm = @algorithm, integrity_key_id = @key_id, integrity_tag = @tag
                           WHERE actor_id = @actor AND tenant_id = @tenant;
                           """;
        await using var command = new NpgsqlCommand(sql, connection, transaction);
        AddParameters(command, actorId, tenantId, allowed, version, fingerprint);
        command.Parameters.AddWithValue("algorithm", AuthorizationContextIntegrityKeyRing.CurrentAlgorithmVersion);
        command.Parameters.AddWithValue("key_id", Integrity.ActiveKeyId);
        command.Parameters.AddWithValue("tag",
            Integrity.ComputeContextTag(actorId, tenantId, allowed, version, fingerprint));
        if (await command.ExecuteNonQueryAsync(cancellationToken) != 1)
            throw new InvalidOperationException("Authorization context update did not affect exactly one row.");
    }

    public Task RevokeAsync(
        NpgsqlConnection connection, NpgsqlTransaction transaction, Guid actorId, int tenantId,
        long version, string fingerprint, CancellationToken cancellationToken = default) =>
        UpdateAsync(connection, transaction, actorId, tenantId, false, version, fingerprint, cancellationToken);

    public Task DeleteAsync(
        NpgsqlConnection connection, NpgsqlTransaction transaction, Guid actorId, int tenantId,
        CancellationToken cancellationToken = default) =>
        DeleteAsync(connection, transaction, actorId, tenantId,
            new AuthorizationWriteProvenance(Guid.Empty, actorId, tenantId, 0, ""), cancellationToken);

    public async Task DeleteAsync(
        NpgsqlConnection connection, NpgsqlTransaction transaction, Guid actorId, int tenantId,
        AuthorizationWriteProvenance provenance, CancellationToken cancellationToken = default)
    {
        await ValidateWriterAsync(connection, transaction, provenance, actorId, tenantId, cancellationToken);
        var current = await LoadForUpdateAsync(connection, transaction, actorId, tenantId, cancellationToken);
        if (current is null) return;

        const string tombstoneSql = """
                                    INSERT INTO banking.authorization_context_tombstone(actor_id, tenant_id, last_version, last_fingerprint, integrity_algorithm, integrity_key_id, integrity_tag)
                                    VALUES (@actor, @tenant, @version, @fingerprint, @algorithm, @key_id, @tag)
                                    ON CONFLICT (actor_id, tenant_id) DO UPDATE SET
                                        last_version = EXCLUDED.last_version, last_fingerprint = EXCLUDED.last_fingerprint,
                                        integrity_algorithm = EXCLUDED.integrity_algorithm, integrity_key_id = EXCLUDED.integrity_key_id, integrity_tag = EXCLUDED.integrity_tag;
                                    """;
        await using (var tombstone = new NpgsqlCommand(tombstoneSql, connection, transaction))
        {
            tombstone.Parameters.AddWithValue("actor", actorId);
            tombstone.Parameters.AddWithValue("tenant", tenantId);
            tombstone.Parameters.AddWithValue("version", current.Version);
            tombstone.Parameters.AddWithValue("fingerprint", current.Fingerprint);
            tombstone.Parameters.AddWithValue("algorithm",
                AuthorizationContextIntegrityKeyRing.CurrentAlgorithmVersion);
            tombstone.Parameters.AddWithValue("key_id", Integrity.ActiveKeyId);
            tombstone.Parameters.AddWithValue("tag",
                Integrity.ComputeTombstoneTag(actorId, tenantId, current.Version, current.Fingerprint));
            await tombstone.ExecuteNonQueryAsync(cancellationToken);
        }

        const string sql = "DELETE FROM banking.authorization_context WHERE actor_id = @actor AND tenant_id = @tenant;";
        await using var command = new NpgsqlCommand(sql, connection, transaction);
        command.Parameters.AddWithValue("actor", actorId);
        command.Parameters.AddWithValue("tenant", tenantId);
        if (await command.ExecuteNonQueryAsync(cancellationToken) != 1)
            throw new InvalidOperationException("Authorization context deletion did not affect exactly one row.");
    }

    private static async Task ValidateWriterAsync(
        NpgsqlConnection connection, NpgsqlTransaction transaction,
        AuthorizationWriteProvenance provenance, Guid actorId, int tenantId,
        CancellationToken cancellationToken)
    {
        if (provenance.WriterId == Guid.Empty)
            throw new UnauthorizedAccessException("Authorization context writes require registered writer provenance.");
        if (provenance.ActorId != actorId || provenance.TenantId != tenantId)
            throw new UnauthorizedAccessException(
                "Authorization writer is not scoped to the target authorization identity.");
        if (provenance.WriteSequence <= 0)
            throw new ArgumentOutOfRangeException(nameof(provenance), "Writer sequence must be positive.");
        if (string.IsNullOrWhiteSpace(provenance.CredentialFingerprint))
            throw new ArgumentException("Writer credential fingerprint is required.", nameof(provenance));

        const string sql = """
                           SELECT tenant_id, actor_id, active, database_role, credential_fingerprint, last_write_sequence
                           FROM banking.authorization_context_writer
                           WHERE writer_id = @writer
                           FOR UPDATE;
                           """;
        await using var command = new NpgsqlCommand(sql, connection, transaction);
        command.Parameters.AddWithValue("writer", provenance.WriterId);
        await using var reader = await command.ExecuteReaderAsync(cancellationToken);
        if (!await reader.ReadAsync(cancellationToken))
            throw new UnauthorizedAccessException("Authorization writer is not registered.");

        var writerTenant = reader.GetInt32(0);
        var writerActor = reader.GetGuid(1);
        var active = reader.GetBoolean(2);
        var databaseRole = reader.GetString(3);
        var credentialFingerprint = reader.GetString(4);
        var lastSequence = reader.GetInt64(5);

        if (!active || writerTenant != tenantId || writerActor != actorId)
            throw new UnauthorizedAccessException(
                "Authorization writer is inactive or outside its tenant/actor scope.");
        if (!string.Equals(credentialFingerprint, provenance.CredentialFingerprint, StringComparison.Ordinal))
            throw new UnauthorizedAccessException(
                "Authorization writer credential provenance does not match the registered credential.");
        if (!string.Equals(databaseRole, await CurrentDatabaseRoleAsync(connection, transaction, cancellationToken),
                StringComparison.Ordinal))
            throw new UnauthorizedAccessException(
                "Authorization writer database-role provenance does not match the current PostgreSQL session.");
        if (provenance.WriteSequence <= lastSequence)
            throw new InvalidOperationException(
                $"Authorization writer sequence is stale. Last={lastSequence}, requested={provenance.WriteSequence}.");
        await reader.DisposeAsync();

        const string update =
            "UPDATE banking.authorization_context_writer SET last_write_sequence = @sequence WHERE writer_id = @writer;";
        await using var updateCommand = new NpgsqlCommand(update, connection, transaction);
        updateCommand.Parameters.AddWithValue("writer", provenance.WriterId);
        updateCommand.Parameters.AddWithValue("sequence", provenance.WriteSequence);
        if (await updateCommand.ExecuteNonQueryAsync(cancellationToken) != 1)
            throw new InvalidOperationException("Authorization writer sequence update did not affect exactly one row.");
    }

    private static async Task<string> CurrentDatabaseRoleAsync(NpgsqlConnection connection,
        NpgsqlTransaction transaction, CancellationToken cancellationToken)
    {
        await using var command = new NpgsqlCommand("SELECT current_user;", connection, transaction);
        return (string)(await command.ExecuteScalarAsync(cancellationToken))!;
    }

    private static void AddParameters(NpgsqlCommand command, Guid actorId, int tenantId, bool allowed, long version,
        string fingerprint)
    {
        command.Parameters.AddWithValue("actor", actorId);
        command.Parameters.AddWithValue("tenant", tenantId);
        command.Parameters.AddWithValue("allowed", allowed);
        command.Parameters.AddWithValue("version", version);
        command.Parameters.AddWithValue("fingerprint", fingerprint);
    }

    private static void ValidateVersion(long version)
    {
        if (version <= 0)
            throw new ArgumentOutOfRangeException(nameof(version), "Authorization context versions must be positive.");
    }

    private static void ValidateFingerprint(string fingerprint)
    {
        if (string.IsNullOrWhiteSpace(fingerprint))
            throw new ArgumentException("Authorization context fingerprint cannot be empty.", nameof(fingerprint));
    }
}

public sealed record AuthorizationContextRow(
    Guid ActorId,
    int TenantId,
    bool Allowed,
    long Version,
    string Fingerprint,
    string IntegrityAlgorithm,
    string IntegrityKeyId,
    string IntegrityTag);

public sealed record AuthorizationWriteProvenance(
    Guid WriterId,
    Guid ActorId,
    int TenantId,
    long WriteSequence,
    string CredentialFingerprint);

File: src/java/benchmarks/foundgine-agent/src/main/java/com/foundgine/benchmarks/highassurance/postgres/PostgresAuthorizationContextStore.java
package com.foundgine.benchmarks.highassurance.postgres;

import com.foundgine.runtime.controlplane.AuthorizationContextIntegrityKeyRing;
import com.foundgine.runtime.controlplane.AuthorizationContextIntegrityKeyRingManager;

import java.sql.*;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;

/**
 * PostgreSQL-backed authorization-context lifecycle boundary.
 *
 * <p>This is the Java parity port of the C# high-assurance store. Integrity is verified before
 * authorization evidence is trusted; lifecycle writes require registered writer provenance and a
 * strictly increasing writer sequence.
 */
public final class PostgresAuthorizationContextStore {
    public record AuthorizationContextRow(
            UUID actorId,
            int tenantId,
            boolean allowed,
            long version,
            String fingerprint,
            String integrityAlgorithm,
            String integrityKeyId,
            String integrityTag) {}

    public record AuthorizationWriteProvenance(
            UUID writerId,
            UUID actorId,
            int tenantId,
            long writeSequence,
            String credentialFingerprint) {}

    private final AuthorizationContextIntegrityKeyRingManager integrityManager;

    public PostgresAuthorizationContextStore(AuthorizationContextIntegrityKeyRing integrity) {
        this(new AuthorizationContextIntegrityKeyRingManager(Objects.requireNonNull(integrity)));
    }

    public PostgresAuthorizationContextStore(
            AuthorizationContextIntegrityKeyRingManager integrityManager) {
        this.integrityManager = Objects.requireNonNull(integrityManager);
    }

    public AuthorizationContextIntegrityKeyRing currentIntegrityKeyRing() {
        return integrityManager.snapshot();
    }

    public Set<String> getPersistedIntegrityKeyIds(Connection connection, boolean transactional)
            throws SQLException {
        var result = new HashSet<String>();
        try (var command =
                connection.prepareStatement(
                        "SELECT integrity_key_id FROM banking.authorization_context UNION SELECT"
                            + " integrity_key_id FROM banking.authorization_context_tombstone")) {
            try (var reader = command.executeQuery()) {
                while (reader.next()) result.add(reader.getString(1));
            }
        }
        return Set.copyOf(result);
    }

    public AuthorizationContextRow loadForUpdate(Connection connection, UUID actorId, int tenantId)
            throws SQLException {
        String sql =
                "SELECT actor_id, tenant_id, allowed, version, fingerprint, integrity_algorithm,"
                    + " integrity_key_id, integrity_tag FROM banking.authorization_context WHERE"
                    + " actor_id=? AND tenant_id=? FOR UPDATE";
        try (var command = connection.prepareStatement(sql)) {
            command.setObject(1, actorId);
            command.setInt(2, tenantId);
            try (var reader = command.executeQuery()) {
                if (!reader.next()) return null;
                var row =
                        new AuthorizationContextRow(
                                reader.getObject(1, UUID.class),
                                reader.getInt(2),
                                reader.getBoolean(3),
                                reader.getLong(4),
                                reader.getString(5),
                                reader.getString(6),
                                reader.getString(7),
                                reader.getString(8));
                if (!integrityManager
                        .snapshot()
                        .verifyContextTag(
                                row.actorId(),
                                row.tenantId(),
                                row.allowed(),
                                row.version(),
                                row.fingerprint(),
                                row.integrityAlgorithm(),
                                row.integrityKeyId(),
                                row.integrityTag())) {
                    throw new IllegalStateException(
                            "Authorization context integrity verification failed; authorization"
                                    + " fails closed.");
                }
                return row;
            }
        }
    }

    public void create(
            Connection connection,
            UUID actorId,
            int tenantId,
            boolean allowed,
            long version,
            String fingerprint,
            AuthorizationWriteProvenance provenance)
            throws SQLException {
        validate(version, fingerprint);
        validateWriter(connection, provenance, actorId, tenantId);
        verifyTombstoneAllowsCreate(connection, actorId, tenantId, version);

        var ring = integrityManager.snapshot();
        String sql =
                "INSERT INTO banking.authorization_context"
                    + " (actor_id,tenant_id,allowed,version,fingerprint,integrity_algorithm,integrity_key_id,integrity_tag)"
                    + " VALUES (?,?,?,?,?,?,?,?)";
        try (var command = connection.prepareStatement(sql)) {
            bind(command, actorId, tenantId, allowed, version, fingerprint);
            command.setString(6, AuthorizationContextIntegrityKeyRing.currentAlgorithmVersion());
            command.setString(7, ring.activeKeyId());
            command.setString(
                    8, ring.computeContextTag(actorId, tenantId, allowed, version, fingerprint));
            if (command.executeUpdate() != 1)
                throw new IllegalStateException(
                        "Authorization context create did not affect exactly one row.");
        }
    }

    public void update(
            Connection connection,
            UUID actorId,
            int tenantId,
            boolean allowed,
            long version,
            String fingerprint,
            AuthorizationWriteProvenance provenance)
            throws SQLException {
        validate(version, fingerprint);
        validateWriter(connection, provenance, actorId, tenantId);
        var current = loadForUpdate(connection, actorId, tenantId);
        if (current == null)
            throw new IllegalStateException("Authorization context does not exist.");
        if (version <= current.version())
            throw new IllegalStateException(
                    "Authorization context version must increase monotonically.");

        var ring = integrityManager.snapshot();
        String sql =
                "UPDATE banking.authorization_context SET"
                    + " allowed=?,version=?,fingerprint=?,integrity_algorithm=?,integrity_key_id=?,integrity_tag=?"
                    + " WHERE actor_id=? AND tenant_id=?";
        try (var command = connection.prepareStatement(sql)) {
            command.setBoolean(1, allowed);
            command.setLong(2, version);
            command.setString(3, fingerprint);
            command.setString(4, AuthorizationContextIntegrityKeyRing.currentAlgorithmVersion());
            command.setString(5, ring.activeKeyId());
            command.setString(
                    6, ring.computeContextTag(actorId, tenantId, allowed, version, fingerprint));
            command.setObject(7, actorId);
            command.setInt(8, tenantId);
            if (command.executeUpdate() != 1)
                throw new IllegalStateException(
                        "Authorization context update did not affect exactly one row.");
        }
    }

    public void revoke(
            Connection connection,
            UUID actorId,
            int tenantId,
            long version,
            String fingerprint,
            AuthorizationWriteProvenance provenance)
            throws SQLException {
        update(connection, actorId, tenantId, false, version, fingerprint, provenance);
    }

    public void delete(
            Connection connection,
            UUID actorId,
            int tenantId,
            AuthorizationWriteProvenance provenance)
            throws SQLException {
        validateWriter(connection, provenance, actorId, tenantId);
        var current = loadForUpdate(connection, actorId, tenantId);
        if (current == null) return;

        var ring = integrityManager.snapshot();
        String tombstone =
                "INSERT INTO banking.authorization_context_tombstone"
                    + " (actor_id,tenant_id,last_version,last_fingerprint,integrity_algorithm,integrity_key_id,integrity_tag)"
                    + " VALUES (?,?,?,?,?,?,?) ON CONFLICT (actor_id,tenant_id) DO UPDATE SET "
                    + "last_version=EXCLUDED.last_version,last_fingerprint=EXCLUDED.last_fingerprint,integrity_algorithm=EXCLUDED.integrity_algorithm,integrity_key_id=EXCLUDED.integrity_key_id,integrity_tag=EXCLUDED.integrity_tag";
        try (var command = connection.prepareStatement(tombstone)) {
            command.setObject(1, actorId);
            command.setInt(2, tenantId);
            command.setLong(3, current.version());
            command.setString(4, current.fingerprint());
            command.setString(5, AuthorizationContextIntegrityKeyRing.currentAlgorithmVersion());
            command.setString(6, ring.activeKeyId());
            command.setString(
                    7,
                    ring.computeTombstoneTag(
                            actorId, tenantId, current.version(), current.fingerprint()));
            command.executeUpdate();
        }
        try (var command =
                connection.prepareStatement(
                        "DELETE FROM banking.authorization_context WHERE actor_id=? AND"
                                + " tenant_id=?")) {
            command.setObject(1, actorId);
            command.setInt(2, tenantId);
            if (command.executeUpdate() != 1)
                throw new IllegalStateException(
                        "Authorization context deletion did not affect exactly one row.");
        }
    }

    private void verifyTombstoneAllowsCreate(
            Connection connection, UUID actorId, int tenantId, long version) throws SQLException {
        String sql =
                "SELECT"
                    + " last_version,last_fingerprint,integrity_algorithm,integrity_key_id,integrity_tag"
                    + " FROM banking.authorization_context_tombstone WHERE actor_id=? AND"
                    + " tenant_id=? FOR UPDATE";
        try (var command = connection.prepareStatement(sql)) {
            command.setObject(1, actorId);
            command.setInt(2, tenantId);
            try (var reader = command.executeQuery()) {
                if (!reader.next()) return;
                var lastVersion = reader.getLong(1);
                var ring = integrityManager.snapshot();
                if (!ring.verifyTombstoneTag(
                        actorId,
                        tenantId,
                        lastVersion,
                        reader.getString(2),
                        reader.getString(3),
                        reader.getString(4),
                        reader.getString(5)))
                    throw new IllegalStateException(
                            "Authorization lifecycle tombstone integrity verification failed;"
                                    + " authorization fails closed.");
                if (version <= lastVersion)
                    throw new IllegalStateException(
                            "Authorization context version must exceed the last committed lifecycle"
                                    + " version.");
            }
        }
    }

    private static void validate(long version, String fingerprint) {
        if (version <= 0)
            throw new IllegalArgumentException("Authorization context versions must be positive.");
        if (fingerprint == null || fingerprint.isBlank())
            throw new IllegalArgumentException(
                    "Authorization context fingerprint cannot be empty.");
    }

    private static void validateWriter(
            Connection connection, AuthorizationWriteProvenance p, UUID actorId, int tenantId)
            throws SQLException {
        if (p == null || p.writerId() == null || p.writerId().equals(new UUID(0, 0)))
            throw new SecurityException(
                    "Authorization context writes require registered writer provenance.");
        if (!actorId.equals(p.actorId()) || tenantId != p.tenantId())
            throw new SecurityException(
                    "Authorization writer is not scoped to the target authorization identity.");
        if (p.writeSequence() <= 0)
            throw new IllegalArgumentException("Writer sequence must be positive.");
        if (p.credentialFingerprint() == null || p.credentialFingerprint().isBlank())
            throw new IllegalArgumentException("Writer credential fingerprint is required.");

        String sql =
                "SELECT"
                    + " tenant_id,actor_id,active,database_role,credential_fingerprint,last_write_sequence"
                    + " FROM banking.authorization_context_writer WHERE writer_id=? FOR UPDATE";
        try (var command = connection.prepareStatement(sql)) {
            command.setObject(1, p.writerId());
            try (var reader = command.executeQuery()) {
                if (!reader.next())
                    throw new SecurityException("Authorization writer is not registered.");
                if (!reader.getBoolean(3)
                        || reader.getInt(1) != tenantId
                        || !actorId.equals(reader.getObject(2, UUID.class)))
                    throw new SecurityException(
                            "Authorization writer is inactive or outside its tenant/actor scope.");
                if (!reader.getString(5).equals(p.credentialFingerprint()))
                    throw new SecurityException(
                            "Authorization writer credential provenance does not match the"
                                    + " registered credential.");
                if (!reader.getString(4).equals(currentDatabaseRole(connection)))
                    throw new SecurityException(
                            "Authorization writer database-role provenance does not match the"
                                    + " current PostgreSQL session.");
                if (p.writeSequence() <= reader.getLong(6))
                    throw new IllegalStateException("Authorization writer sequence is stale.");
            }
        }
        try (var command =
                connection.prepareStatement(
                        "UPDATE banking.authorization_context_writer SET last_write_sequence=?"
                                + " WHERE writer_id=?")) {
            command.setLong(1, p.writeSequence());
            command.setObject(2, p.writerId());
            if (command.executeUpdate() != 1)
                throw new IllegalStateException(
                        "Authorization writer sequence update did not affect exactly one row.");
        }
    }

    private static String currentDatabaseRole(Connection connection) throws SQLException {
        try (var command = connection.prepareStatement("SELECT current_user");
                var reader = command.executeQuery()) {
            if (!reader.next())
                throw new IllegalStateException("Unable to resolve PostgreSQL current_user.");
            return reader.getString(1);
        }
    }

    private static void bind(
            PreparedStatement command,
            UUID actorId,
            int tenantId,
            boolean allowed,
            long version,
            String fingerprint)
            throws SQLException {
        command.setObject(1, actorId);
        command.setInt(2, tenantId);
        command.setBoolean(3, allowed);
        command.setLong(4, version);
        command.setString(5, fingerprint);
    }
}
